### PR TITLE
Inject cluster name as an environment variable into head and worker pods

### LINKED ISSUE #928: [Feature] Inject cluster name as an environment variable into head and worker pods
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/kuberay/issues) and found no similar feature requirement.

### Description

The cluster name should be easily available from inside Kubernetes pods inside the cluster. An env var would be more convenient than e.g. a Ray API call, so that this would be available both from inside Python and from e.g. bash scripts.

### Use case

I'd like to use the cluster name for namespacing extra information I'll be storing in an external Redis.

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by Yicheng-Lu-llll**
## Why are these changes needed?

See https://github.com/ray-project/kuberay/issues/928.

Now user can get env var `RAY_CLUSTER_NAME` in head and worker pods.

## Related issue number
Closes #928

## Checks
```shell
rootPath={Your kuberay repo path}
# ray-cluster.complete.yaml will create one head pod and one worker pod.
kubectl apply -f $rootPath/ray-operator/config/samples/ray-cluster.complete.yaml
kubectl exec -it $(kubectl get pods -o=name | grep head) -- env | grep RAY_CLUSTER_NAME
kubectl exec -it $(kubectl get pods -o=name | grep worker) -- env | grep RAY_CLUSTER_NAME
# Now change the pod replicas to 2, kuberay operator will create a new worker pod, check if the newly created pod also have env vars.
kubectl apply -f $rootPath/ray-operator/config/samples/ray-cluster.complete.yaml
kubectl exec -it raycluster-complete-worker-small-group-fw6kk -- env | grep RAY_CLUSTER_NAME


# will always see:
# RAY_CLUSTER_NAME=raycluster-complete

```

### CODE REVIEW

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 1]
```diff
@@ -799,7 +799,7 @@ func (r *RayClusterReconciler) buildWorkerPod(instance rayiov1alpha1.RayCluster,
 	autoscalingEnabled := instance.Spec.EnableInTreeAutoscaling
 	podTemplateSpec := common.DefaultWorkerPodTemplate(instance, worker, podName, svcName, headPort)
 	creatorName := getCreator(instance)
-	pod := common.BuildPod(podTemplateSpec, rayiov1alpha1.WorkerNode, worker.RayStartParams, svcName, headPort, autoscalingEnabled, creatorName)
+	pod := common.BuildPod(podTemplateSpec, rayiov1alpha1.WorkerNode, worker.RayStartParams, svcName, headPort, autoscalingEnabled, creatorName, instance.Name)
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:802]:
```go
podName := strings.ToLower(instance.Name + common.DashSymbol + string(rayiov1alpha1.WorkerNode) + common.DashSymbol + worker.GroupName + common.DashSymbol)
podName = utils.CheckName(podName) // making sure the name is valid
svcName := utils.GenerateServiceName(instance.Name)
svcName = utils.CheckName(svcName)
```

There seems to be a lot of duplicate information for `BuildPod`. For example, `svcName` is derived from `instance.Name`, and this PR passes `instance.Name` into `BuildPod` again.

In addition, can you check whether `utils.CheckName` will always be called after podName and svcName. If so, we need to refactor the logic.

**Review by kevin85421 — COMMENTED** [review]
Thank you for your contribution! Could you 

(1) Think about the comparisons between injecting the env variable by KubeRay operator or setting by YAML files.

(2) Add some tests.

(3) Test autoscaler.

**Review by kevin85421 — COMMENTED** [review]
Check whether `BuildPod` will always be called after `DefaultWorkerPodTemplate` and `DefaultHeadPodTemplate`.

File: ray-operator/controllers/ray/common/pod.go [hunk 2]
```diff
@@ -573,6 +573,10 @@ func setContainerEnvVars(pod *v1.Pod, rayContainerIndex int, rayNodeType rayiov1
 		portEnv := v1.EnvVar{Name: RAY_PORT, Value: headPort}
 		container.Env = append(container.Env, portEnv)
 	}
+	if !envVarExists(CLUSTER_NAME, container.Env) {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/common/pod.go:576]: `KUBERAY_CR_NAME`?

**Comment by kevin85421** [comment]
This looks like a better solution?

https://github.com/ray-project/kuberay/blob/master/ray-operator/controllers/ray/common/pod.go#L386

**Comment by Yicheng-Lu-llll** [comment]
> This looks like a better solution?
> 
> https://github.com/ray-project/kuberay/blob/master/ray-operator/controllers/ray/common/pod.go#L386

Updated. Thank you for pointing that out!

File: ray-operator/controllers/ray/common/constant.go [hunk 3]
```diff
@@ -72,7 +72,7 @@ const (
 
 	// Use as container env variable
 	NAMESPACE                               = "NAMESPACE"
-	CLUSTER_NAME                            = "CLUSTER_NAME"
+	RAY_CLUSTER_NAME                        = "RAY_CLUSTER_NAME"
```
Comment (kevin85421) [inline ray-operator/controllers/ray/common/constant.go:75]: `KUBERAY_RAYCLUSTER_NAME` or `KUBERAY_CLUSTER_NAME`? 

It is very easy for Ray community to use the environment variable with the name `RAY_CLUSTER_NAME`.

File: ray-operator/controllers/ray/common/pod.go [hunk 4]
```diff
@@ -572,6 +572,19 @@ func setContainerEnvVars(pod *v1.Pod, rayContainerIndex int, rayNodeType rayiov1
 		portEnv := v1.EnvVar{Name: RAY_PORT, Value: headPort}
 		container.Env = append(container.Env, portEnv)
 	}
+
+	if !envVarExists(RAY_CLUSTER_NAME, container.Env) {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/common/pod.go:576]: In my opinion, this environment variable should always be maintained by KubeRay. Users should not be able to overwrite it.

File: ray-operator/controllers/ray/common/pod.go [hunk 5]
```diff
@@ -383,7 +383,7 @@ func BuildAutoscalerContainer(autoscalerImage string) v1.Container {
 				Name: "RAY_CLUSTER_NAME",
```
Comment (kevin85421) [inline ray-operator/controllers/ray/common/pod.go:383]: Use variable `RAY_CLUSTER_NAME`.

File: ray-operator/controllers/ray/common/pod_test.go [hunk 6]
```diff
@@ -326,9 +326,18 @@ func checkPodEnv(t *testing.T, pod v1.Pod, envName string, expectedValue string)
...
 		if env.Name == envName {
-			if !(env.Value == expectedValue) {
-				t.Fatalf("Expected `%v` but got `%v`", expectedValue, env.Value)
+			// env.ValueFrom is the source for the environment variable's value. It will be nil if env.Value is not empty.
+			if env.ValueFrom == nil {
+				if !(env.Value == expectedValue) {
+					t.Fatalf("Expected `%v` but got `%v`", expectedValue, env.Value)
+				}
+			} else {
+				fmt.Println(env.Name, env.Value, env.ValueFrom.FieldRef.FieldPath)
```
Comment (kevin85421) [inline ray-operator/controllers/ray/common/pod_test.go:335]: Remove this line.

File: ray-operator/controllers/ray/common/pod_test.go [hunk 7]
```diff
@@ -326,9 +326,18 @@ func checkPodEnv(t *testing.T, pod v1.Pod, envName string, expectedValue string)
 	foundEnv := false
 	for _, env := range pod.Spec.Containers[0].Env {
 		if env.Name == envName {
-			if !(env.Value == expectedValue) {
-				t.Fatalf("Expected `%v` but got `%v`", expectedValue, env.Value)
+			// env.ValueFrom is the source for the environment variable's value. It will be nil if env.Value is not empty.
```
Comment (kevin85421) [inline ray-operator/controllers/ray/common/pod_test.go:329]: Can you provide the reference?

File: ray-operator/controllers/ray/common/pod_test.go [hunk 8]
```diff
@@ -326,9 +326,18 @@ func checkPodEnv(t *testing.T, pod v1.Pod, envName string, expectedValue string)
...
+			if env.ValueFrom == nil {
+				if !(env.Value == expectedValue) {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/common/pod_test.go:331]:
```suggestion
				if (env.Value != expectedValue) {
```

File: ray-operator/controllers/ray/common/pod_test.go [hunk 9]
```diff
@@ -326,9 +326,18 @@ func checkPodEnv(t *testing.T, pod v1.Pod, envName string, expectedValue string)
...
+					t.Fatalf("Expected `%v` but got `%v`", expectedValue, env.Value)
+				}
+			} else {
+				fmt.Println(env.Name, env.Value, env.ValueFrom.FieldRef.FieldPath)
+				if !(env.ValueFrom.FieldRef.FieldPath == expectedValue) {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/common/pod_test.go:336]:
```suggestion
				if (env.ValueFrom.FieldRef.FieldPath != expectedValue) {
```

**Review by kevin85421 — COMMENTED** [review]
Can you add a document for environment_variables? Thanks!

**Review by kevin85421 — COMMENTED** [review]
Should we also update this line?

https://github.com/ray-project/kuberay/blob/1a8895e1d541d15ea0b0f5cb1babbe0cdd9e0ff4/ray-operator/controllers/ray/common/pod_test.go#L194

**Review by kevin85421 — APPROVED** [review]
LGTM. I also tested this PR manually. 

Screen Shot 2023-03-13 at 4 53 17 PM

**Pull request merged**